records/generator.go

```diff
@@ -443,19 +434,25 @@ func (rg *RecordGenerator) taskContextRecord(ctx context, task state.Task, f sta
 	canonical := ctx.taskName + "-" + ctx.taskID + "-" + ctx.slaveID + "." + fname
 	arec := ctx.taskName + "." + fname
 
-	ipv4, ipv6 := splitIPsToTypes(ctx.taskIPs)
-	if ipv4 != nil {
-		rg.insertTaskRR(arec+tail, ipv4.String(), A, enumTask)
-		rg.insertTaskRR(canonical+tail, ipv4.String(), A, enumTask)
-	}
-	if ipv6 != nil {
-		rg.insertTaskRR(arec+tail, ipv6.String(), AAAA, enumTask)
-		rg.insertTaskRR(canonical+tail, ipv6.String(), AAAA, enumTask)
+	// Only use the first ipv4 and first ipv6 found in sources
+	tIPs := ipsTo4And6(ctx.taskIPs)
+	for _, tIP := range tIPs {
+		rg.insertTaskRR(arec+tail, tIP.String(), rrsKindForIP(tIP), enumTask)
+		rg.insertTaskRR(canonical+tail, tIP.String(), rrsKindForIP(tIP), enumTask)
 	}
 
-	slaveIPKind := rrsKindForIPStr(ctx.slaveIP)
-	rg.insertTaskRR(arec+".slave"+tail, ctx.slaveIP, slaveIPKind, enumTask)
-	rg.insertTaskRR(canonical+".slave"+tail, ctx.slaveIP, slaveIPKind, enumTask)
+	// slaveIPs already only has at most one ipv4 and one ipv6
+	for _, sIPStr := range ctx.slaveIPs {
+		if sIP := net.ParseIP(sIPStr); sIP != nil {
+			rg.insertTaskRR(arec+".slave"+tail, sIP.String(), rrsKindForIP(sIP), enumTask)
+			rg.insertTaskRR(canonical+".slave"+tail, sIP.String(), rrsKindForIP(sIP), enumTask)
+		} else {
+			// ack: slave IP may not be an actual IP if labels.DomainFrag was used.
```

Comment:
good catch. this is suspect. if not debugging/fixing this issue here, should probably create an issue to track

Reply:
Created https://github.com/mesosphere/mesos-dns/issues/509 and I'll reference from the comment.